src/lib.rs

```diff
@@ -225,9 +225,14 @@ impl_sealed_for!(bool, u8, u16, u32, u64, u128);
 impl PopBits for u8 {
     #[inline(always)]
     fn pop_bits(&mut self, amount: u32) -> u8 {
+        let orig_bits = self.count_ones();
         debug_assert!(0 < amount && amount <= 8);
-        let res = *self & ((0x1_u16.wrapping_shl(amount) as u8).wrapping_sub(1));
-        *self /= 0x80 >> (8 - amount);
+        let res = *self & ((0x1_u16.wrapping_shl(amount)).wrapping_sub(1) as u8);
+        *self = match self.overflowing_shr(amount) {
+            (v, false) => v,
+            _ => 0,
+        };
+        debug_assert_eq!(res.count_ones() + self.count_ones(), orig_bits);
```

Comment:
I was aware of complications due to shift operations since they do not behave as I would want them to do (just what you did here). And back then I tried really hard to achieve the same semantics with pure computation and hoped it worked. Seems that I was a bit off here. I am a bit worried about performance though since Rust has some problems with trait methods and `const fn` at the moment. Also this code is kind of a "hot path" code for this crate so we should really benchmark the changes first before we accept them.

Could you please do a before-after on your 5 years old system and paste them here?

Reply:
With Changes:
> test verify_benches::get_set_a ... ignored
> test verify_benches::get_set_b ... ignored
> test verify_benches::get_set_c ... ignored
> test verify_benches::get_set_d ... ignored
> test verify_benches::get_set_e ... ignored
> test generated::get_a   ... bench:          77 ns/iter (+/- 3)
> test generated::get_b   ... bench:          77 ns/iter (+/- 3)
> test generated::get_c   ... bench:          77 ns/iter (+/- 1)
> test generated::get_d   ... bench:          77 ns/iter (+/- 3)
> test generated::get_e   ... bench:          77 ns/iter (+/- 4)
> test generated::set_a   ... bench:         388 ns/iter (+/- 18)
> test generated::set_b   ... bench:         420 ns/iter (+/- 21)
> test generated::set_c   ... bench:         421 ns/iter (+/- 11)
> test generated::set_d   ... bench:         423 ns/iter (+/- 17)
> test generated::set_e   ... bench:         145 ns/iter (+/- 5)
> test handwritten::get_a ... bench:          77 ns/iter (+/- 4)
> test handwritten::get_b ... bench:          76 ns/iter (+/- 2)
> test handwritten::get_c ... bench:          78 ns/iter (+/- 4)
> test handwritten::get_d ... bench:          77 ns/iter (+/- 3)
> test handwritten::get_e ... bench:          77 ns/iter (+/- 3)
> test handwritten::set_a ... bench:         475 ns/iter (+/- 8)
> test handwritten::set_b ... bench:         486 ns/iter (+/- 10)
> test handwritten::set_c ... bench:         418 ns/iter (+/- 9)
> test handwritten::set_d ... bench:         487 ns/iter (+/- 12)
> test handwritten::set_e ... bench:         145 ns/iter (+/- 4)

Prior to Changes:
> test verify_benches::get_set_a ... ignored
> test verify_benches::get_set_b ... ignored
> test verify_benches::get_set_c ... ignored
> test verify_benches::get_set_d ... ignored
> test verify_benches::get_set_e ... ignored
> test generated::get_a   ... bench:          77 ns/iter (+/- 3)
> test generated::get_b   ... bench:          77 ns/iter (+/- 3)
> test generated::get_c   ... bench:          77 ns/iter (+/- 3)
> test generated::get_d   ... bench:          77 ns/iter (+/- 3)
> test generated::get_e   ... bench:          77 ns/iter (+/- 3)
> test generated::set_a   ... bench:         388 ns/iter (+/- 19)
> test generated::set_b   ... bench:         420 ns/iter (+/- 21)
> test generated::set_c   ... bench:         423 ns/iter (+/- 11)
> test generated::set_d   ... bench:         419 ns/iter (+/- 14)
> test generated::set_e   ... bench:         145 ns/iter (+/- 3)
> test handwritten::get_a ... bench:          77 ns/iter (+/- 3)
> test handwritten::get_b ... bench:          77 ns/iter (+/- 6)
> test handwritten::get_c ... bench:          77 ns/iter (+/- 2)
> test handwritten::get_d ... bench:          77 ns/iter (+/- 3)
> test handwritten::get_e ... bench:          78 ns/iter (+/- 6)
> test handwritten::set_a ... bench:         475 ns/iter (+/- 20)
> test handwritten::set_b ... bench:         484 ns/iter (+/- 16)
> test handwritten::set_c ... bench:         415 ns/iter (+/- 13)
> test handwritten::set_d ... bench:         484 ns/iter (+/- 14)
> test handwritten::set_e ... bench:         145 ns/iter (+/- 5)